### PR TITLE
Kotlin collections

### PR DESCRIPTION
**Pull request by SHeimlich**
Includes updates to string encoder/decoder to include embedded dictionaries and lists.

Has final updates related to #3900 

Includes Kotlin Tutorial 5 - Collections

### CODE REVIEW

File: src/wasm/kotlin/src/arcs/Entity.kt [hunk 1]
```diff
@@ -12,38 +12,38 @@ class StringDecoder(private var str: String) {
 
     companion object {
 
-      fun decodeDictionary(str: String): Map<String, String> {
-        val decoder = StringDecoder(str)
-        val dict = mutableMapOf<String, String>()
+        fun decodeDictionary(str: String): Map<String, String> {
```
[the remark this answers is not in the document]
Comment (jasonwyatt) [reply src/wasm/kotlin/src/arcs/Entity.kt]: After looking around github.com/google and a bunch of other libraries, it seems that 4-spaces is the way to go afterall.  I'm sorry for the churn on this. 

It might make dealing lint on g3 a pain if people try to modify Arcs kotlin over there in order to contribute back to Github, but I think it's probably for the best to keep Arcs Github in line with how google tends to write open source kotlin.

File: src/runtime/wasm.ts [hunk 2]
```diff
@@ -209,12 +210,32 @@ export class StringDecoder {
       }
       // otherwise typeChar is value-type specifier
       else {
-        dict[key] = this.decodeValue(typeChar);
+        dict[key] = this.decodeValue(typeChar);      
       }
     }
     return dict;
   }
```
Comment (cromwellian) [inline src/runtime/wasm.ts]: I feel like these fixes should be split into a separate PR.

File: particles/Tutorial/Kotlin/5_Collections/Collections.kt [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import arcs.*
+import kotlin.native.internal.ExportForCppRuntime
+
+/**
+ * Sample Kotlin-WASM Particle to use a JSON store.
+ */
+class CollectionsParticle : Particle() {
+
+   private val people = Collection { CollectionsParticle_InputData() }
+   init {
```
Comment (jasonwyatt) [inline particles/Tutorial/Kotlin/5_Collections/Collections.kt]: nit: I would place a blank line between the instance variables and the init block

File: particles/Tutorial/Kotlin/5_Collections/Collections.kt [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+        return model + mapOf(
+            "people" to mapOf(
+                "\$template" to "person",
+                "models" to peopleList
+            )
+        )   
+    }
+
+    override fun getTemplate(slotName: String): String {
```
Comment (jasonwyatt) [inline particles/Tutorial/Kotlin/5_Collections/Collections.kt]: Optional nit: this could be a function expression:

```kotlin
override fun getTemplate(slotName: String): String =
    """
    Hello to everyone:
    <ul>{{people}}</ul>
    
    <template person>
      <!-- ...... -->
    </template>
    """.trimIndent()
```

File: src/wasm/kotlin/src/arcs/Entity.kt [hunk 5]
```diff
@@ -12,38 +12,38 @@ class StringDecoder(private var str: String) {
...
+            val decoder = StringDecoder(str)
+            val dict = mutableMapOf<String, String>()
 
-        var num = decoder.getInt(":")
-        while(num-- > 0){
-          val klen = decoder.getInt(":")
-          val key = decoder.chomp(klen)
+            var num = decoder.getInt(":")
+            while(num-- > 0){
```
Comment (jasonwyatt) [inline src/wasm/kotlin/src/arcs/Entity.kt]: This loop could be done with: `repeat`. E.g:

```kotlin
repeat(getInt(":")) {
    // Inside of loop.
}
```

File: src/wasm/kotlin/src/arcs/Entity.kt [hunk 6]
```diff
@@ -12,38 +12,38 @@ class StringDecoder(private var str: String) {
...
+          val list = mutableListOf<String>()
 
-        var num = decoder.getInt(":")
-        while(num-- > 0) {
-          val len = decoder.getInt(":")
-          val chunk = decoder.chomp(len)
-          list.add(chunk)
-        }
+          var num = decoder.getInt(":")
+            while(num-- > 0) {
```
Comment (jasonwyatt) [inline src/wasm/kotlin/src/arcs/Entity.kt]: This might even be able to be done with `downTo` and `map`:

```kotlin
list += with(decoder) {
    getInt(":").downTo(0).map { chomp(getInt(":")) }
}
```

File: src/wasm/kotlin/src/arcs/Entity.kt [hunk 7]
```diff
@@ -96,16 +96,49 @@ class StringDecoder(private var str: String) {
...
 
     companion object {
-      fun encodeDictionary(dict: Map<String, String>): String {
-        val sb = StringBuilder()
-        sb.append(dict.size).append(":")
+        fun encodeDictionary(dict: Map<String, Any>): String {
+            val sb = StringBuilder()
+            sb.append(dict.size).append(":")
+
+            for((key, value) in dict) {
```
Comment (jasonwyatt) [inline src/wasm/kotlin/src/arcs/Entity.kt]: this could be replaced with `.joinToString`:

```kotlin
sb.append(dict.joinToString(delimeter = "") { (key, value) -> "${key.length}:$key${encodeValue(value)}" })
```

also, I believe `+=` should work with StringBuilder in kotlin, so you can avoid doing `.append`

File: src/wasm/kotlin/src/arcs/Entity.kt [hunk 8]
```diff
@@ -96,16 +96,49 @@ class StringDecoder(private var str: String) {
...
+                sb.append(encodeValue(value))
+            }
+            return sb.toString()
+        }
+
+        fun encodeList(list: List<Any>): String {
+          val sb = StringBuilder()
+          sb.append(list.size).append(":")
+          
+          for(value in list) {
```
Comment (jasonwyatt) [inline src/wasm/kotlin/src/arcs/Entity.kt]: Similarly, `.joinToString` could be used here.

File: src/wasm/kotlin/src/arcs/Entity.kt [hunk 9]
```diff
@@ -96,16 +96,49 @@ class StringDecoder(private var str: String) {
...
+            sb.append(encodeValue(value))
+          }
+          return sb.toString()
+        }
 
-        for((key, value) in dict) {
-          sb.append(key.length).append(":").append(key)
-          sb.append(value.length).append(":").append(value)
+        fun encodeValue(value: Any): String {
```
Comment (jasonwyatt) [inline src/wasm/kotlin/src/arcs/Entity.kt]: Since there aren't any loops within this method, consider using string templates instead of string builder, for readability. (they use StringBuilder under the covers for you)

e.g.

```kotlin
sb.append("T").append(value.length).append(":").append(value)
```
becomes
```kotlin
"T${value.length}:$value"
```

File: src/wasm/kotlin/src/arcs/WasmJsInterop.kt [hunk 10]
```diff
@@ -50,6 +53,11 @@ fun WasmString.toKString(): String {
     return this.toLong().toCPointer<ByteVar>()!!.toKStringFromUtf8()
 }
 
+// Convert a WasmString pointer into a nullable Kotlin String
+fun WasmNullableString.toNullableKString(): String? {
+    return this.toLong().toCPointer<ByteVar>()?.toKStringFromUtf8()
```
Comment (jasonwyatt) [inline src/wasm/kotlin/src/arcs/WasmJsInterop.kt]: Might be tidier as a function expression.

File: src/wasm/kotlin/src/arcs/WasmJsInterop.kt [hunk 11]
```diff
@@ -109,22 +117,22 @@ fun init(particlePtr: WasmAddress) {
 
 @Retain
 @ExportForCppRuntime("_syncHandle")
-fun syncHandle(particlePtr: WasmAddress, handlePtr: WasmAddress, encoded: WasmString) {
+fun syncHandle(particlePtr: WasmAddress, handlePtr: WasmAddress, encoded: WasmNullableString) {
     log("Getting handle")
     val handle = handlePtr.toObject<Handle>()
-    val encodedStr = encoded.toKString()
-    log("Handle is " + handle.name + "syncing '" + encodedStr + "'")
+    val encodedStr: String? = encoded.toNullableKString()
```
Comment (jasonwyatt) [inline src/wasm/kotlin/src/arcs/WasmJsInterop.kt]: Probably don't need the type here, it should be able to be inferred from the result of `toNullableKString` by the compiler.

File: src/wasm/kotlin/src/arcs/WasmJsInterop.kt [hunk 12]
```diff
@@ -109,22 +117,22 @@ fun init(particlePtr: WasmAddress) {
...
     handle.sync(encodedStr)
     log("Invoking sync on handle on particle")
     particlePtr.toObject<Particle>().sync(handle)
 }
 
 @Retain
 @ExportForCppRuntime("_updateHandle")
-fun updateHandle(particlePtr: WasmAddress, handlePtr: WasmAddress, encoded1Ptr: WasmString,
-                 encoded2Ptr: WasmString) {
+fun updateHandle(particlePtr: WasmAddress, handlePtr: WasmAddress, encoded1Ptr: WasmNullableString,
```
Comment (jasonwyatt) [inline src/wasm/kotlin/src/arcs/WasmJsInterop.kt]: Style when arguments need to wrap is to place a newline after the opening `(`.

```kotlin
fun updateHandle(
    particlePtr: WasmAddress, handlePtr: WasmAddress,
    encoded1Ptr: WasmNullableString, encoded2Ptr: WasmNullableString
) {
    // body
}
```

Comment (jasonwyatt) [reply src/wasm/kotlin/src/arcs/WasmJsInterop.kt]: I tend to prefer putting all arguments on separate lines as well in this situation, but that's not a requirement of the style guide.
